xds/src/main/java/io/grpc/xds/ClusterResolverLoadBalancer.java

```diff
@@ -474,31 +476,45 @@ public void run() {
       private LogicalDnsClusterState(String name, @Nullable String lrsServerName,
           @Nullable Long maxConcurrentRequests, @Nullable UpstreamTlsContext tlsContext) {
         super(name, null, lrsServerName, maxConcurrentRequests, tlsContext);
-        NameResolver.Args args = helper.getNameResolverArgs();
-        URI uri;
-        try {
-          uri = new URI(authority);
-        } catch (URISyntaxException e) {
-          // TODO(chengyuanzhang): unlikely to happen, but maybe handle it more gracefully.
-          throw new AssertionError("Bug, invalid authority: " + authority, e);
-        }
-        resolver = helper.getNameResolverRegistry().asFactory().newNameResolver(uri, args);
+        nameResolverFactory =
+            checkNotNull(helper.getNameResolverRegistry().asFactory(), "nameResolverFactory");
+        nameResolverArgs = checkNotNull(helper.getNameResolverArgs(), "nameResolverArgs");
       }
 
       @Override
       void start() {
+        URI uri;
+        try {
+          uri = new URI("dns", "", "/" + authority, null);
```

Comment:
Seems should use the five-arg constructor passing authority directly. cc @ejona86

Reply:
_The authority_ component in a standard URI ([RFC 2396 sec 3.2](https://www.ietf.org/rfc/rfc2396.txt)) isn't the same thing we are referring to as the _Channel authority_ in gRPC. The Channel authority really is the [_path_ in the URI](https://github.com/grpc/grpc-java/blob/42d7fba1b86ad711687944dc91d0ea8ab4908f44/core/src/main/java/io/grpc/internal/ManagedChannelImpl.java#L754). Nothing to do with the five-arg constructor (with query component).

Reply:
The 5-arg constructor mainly lets you pass a query string, which we don't have. The constructor usage here looks reasonable.

> The design is to resolve the original target as if we are not using xDS.

I'm surprised by that. That's not the semantics of STRICT_DNS/LOGICAL_DNS to my knowledge. Those use the `load_assignment` field (or `hosts` earlier in v2) to specify the names to look up (the dns name is buried within the SocketAddress.address of the Endpoint).

Reply:
I was surprised by resolving the original target of the channel as well, but looks this is indeed how TD is going to work for DirectPath. The LOGICAL_DNS Cluster given by the CDS response is just "google_cfe" to tell us to use DNS, it does not give any name to be used for resolving. 

The original target for creating the channel is `google-c2p:///directpath-trafficdirector-testbed.sandbox.googleapis.com`.

Reply:
@markdroth, are we ignoring the `load_assignment` field as required by STATIC_DNS/LOGICAL_DNS?

Reply:
Eric, I think you're misunderstanding how LOGICAL_DNS works.  It can't possibly depend on the endpoints, because we need to know what DNS name to look up in order to get the list of endpoints -- we don't have the endpoints before we do the DNS lookup.

I believe in Envoy, the name it looks up in DNS is the `:authority` (or `Host`) header of the incoming request.  The equivalent for that in gRPC is the server name from the target URI used to create the channel.

Reply:
Thanks for bringing this to my attention -- it definitely looks like we were misunderstanding how this works.  I'll follow up on this separately.